trade-finance-manager-ui/server/controllers/login/index.js

```diff
@@ -1,54 +1,119 @@
+  const errorMessage = `Login url generation failed - please try later or contact us using details below`;
+  return res.status(500).render('_partials/problem-with-service.njk', { error: { message: errorMessage } });
 };
 
-const logout = (req, res) => {
-  req.session.destroy(() => {
-    res.redirect('/');
-  });
+/**
+ * handleSsoRedirect
+ * Because of security restrictions we need to handle 2 POST calls in same URL.
+ * 1) Step 1 - POST is from SSO Authority we don't have session cookie and will prepare local form that will autosubmit.
+ * 2) Step 2 - POST is from our form and now we have session cookie and can verify SSO authority data.
+ * 3) Send SSO authority data to TFM-API for verification and login.
+ * 4) Redirect user to original location, most likely /.
+ * @param {Express.Request} req Express request
+ * @param {Express.Response} res Express response
+ * @returns {Promise<object>} Express response with redirect or error.
+ */
+const handleSsoRedirect = async (req, res) => {
+  try {
+    const { body, session } = req;
+
+    if (!body.formId) {
+      return loginService.acceptExternalSsoPost(req, res);
+    }
+
+    if (!session?.auth?.pkceCodes) {
+      console.error('SSO login session details are missing. Start login from beginning.');
+      return res.redirect('/');
+    }
+
+    const {
+      auth: { pkceCodes, authCodeUrlRequest, authCodeRequest },
+    } = session;
+    const { code, state } = body;
+
+    const apiResponse = await api.processSsoRedirect({ pkceCodes, authCodeUrlRequest, authCodeRequest, code, state });
+
+    if (apiResponse.token) {
+      req.session.userToken = apiResponse.token;
+      req.session.user = apiResponse.tfmUser;
+
+      // Unset data used for SSO validation
+      delete req.session.auth;
+
+      // Unset cookie used for SSO redirect counting
+      res.clearCookie(SSO.REDIRECT_COUNTER.COOKIE_NAME);
+
+      return res.redirect(apiResponse.redirectUrl);
+    }
+    console.error('TFM-UI - login failed in TFM-API, redirect to /');
+    return res.redirect('/');
+  } catch (error) {
+    const errorMessage = error.message || `Login process failed - try again or contact us using details below.`;
+    console.error('TFM-UI - login failed, error %O', error);
+    return res.status(500).render('_partials/problem-with-service.njk', { error: { message: errorMessage } });
+  }
+};
+
+/**
+ * logout
+ * Delete local session and redirect to SSO authority logout page.
+ * 1) Get logout url from TFM-API.
+ * 2) Destroy TFM-UI user login session and redirect to SSO authority logout form.
+ * @param {Express.Request} req Express request
+ * @param {Express.Response} res Express response
+ * @returns {Promise<object>} Express response with redirect or error.
+ */
+const logout = async (req, res) => {
+  const apiResponse = await api.getAuthLogoutUrl(req.session.userToken);
+
+  if (apiResponse.logoutUrl) {
+    return req.session.destroy(() => res.redirect(apiResponse.logoutUrl));
+  }
```

Comment:
I've updated this to not silently fail and redirect, so users are not surprised that they are not logged out